general/mem_manager.cpp

```diff
@@ -77,286 +112,624 @@ MemoryManager::~MemoryManager()
 void* MemoryManager::Insert(void *ptr, const std::size_t bytes)
 {
    if (!UsingMM()) { return ptr; }
-   const bool known = IsKnown(ptr);
-   if (known)
+   if (ptr == NULL)
+   {
+      MFEM_VERIFY(bytes == 0, "Trying to add NULL with size " << bytes);
+      return NULL;
+   }
+   auto res = maps->memories.emplace(ptr, internal::Memory(ptr, bytes));
+   if (res.second == false)
    {
       mfem_error("Trying to add an already present address!");
    }
-   maps->memories.emplace(ptr, internal::Memory(ptr, bytes));
    return ptr;
 }
 
-void *MemoryManager::Erase(void *ptr)
+void *MemoryManager::Erase(void *ptr, bool free_dev_ptr)
 {
    if (!UsingMM()) { return ptr; }
    if (!ptr) { return ptr; }
-   const bool known = IsKnown(ptr);
-   if (!known)
+   auto mem_map_iter = maps->memories.find(ptr);
+   if (mem_map_iter == maps->memories.end())
    {
       mfem_error("Trying to erase an unknown pointer!");
    }
-   internal::Memory &mem = maps->memories.at(ptr);
-   if (mem.d_ptr) { CuMemFree(mem.d_ptr); }
+   internal::Memory &mem = mem_map_iter->second;
+   if (mem.d_ptr && free_dev_ptr) { CuMemFree(mem.d_ptr); }
+#if 0
```

Comment:
Should we remove this before merging?

Reply:
We can. I just left it there while testing.